csharp/src/lesson03/README.md

```diff
@@ -0,0 +1,327 @@
+    .SetTag("http.url", url);       // or: .SetTag(Tags.HttpUrl, url);
+
+var dictionary = new Dictionary<string, string>();
+_tracer.Inject(span.Context, BuiltinFormats.HttpHeaders, new TextMapInjectAdapter(dictionary));
+foreach (var entry in dictionary)
+    _webClient.Headers.Add(entry.Key, entry.Value);
+```
+
+Notice that we also add a couple additional tags to the span with some metadata about the HTTP request,
+and we mark the span with a `span.kind=client` tag, as recommended by the OpenTracing
+[Semantic Conventions][semantic-conventions]. There are other tags we could add. All of the recommended tags 
+are specified in `OpenTracing.Tag.Tags` for easier usage as shown above.
+
+### Instrumenting the Server manually
+
+Our server is currently not instrumented for tracing. We need to do the following:
+
+#### Create an instance of a Tracer, similar to how we did it in `Hello.cs`
+
+Add a tracer property to the `Startup` class:
+
+```csharp
+private static readonly Tracer Tracer = Tracing.Init("Webservice");
+```
+
+Register the tracer and add it to the services, making it available globally through Dependency Injection:
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddMvc();
+
+    services.AddSingleton<ITracer, Tracer>(t => Tracer);
+}
+```
+
+Add a member variable and a constructor to the `FormatController`. The `ITracer` will be supplied using 
+Dependency Injection:
+
+```csharp
+private readonly ITracer _tracer;
+
+public FormatController(ITracer tracer)
+{
+    _tracer = tracer;
+}
+```
+
+#### Extract the span context from the incoming request using `ITracer.Extract`
+
+First, add a helper function:
+
+```csharp
+public static IScope StartServerSpan(ITracer tracer, IDictionary<string, string> headers, string operationName)
+{
+    ISpanBuilder spanBuilder;
+    try
+    {
+        ISpanContext parentSpanCtx = tracer.Extract(BuiltinFormats.HttpHeaders, new TextMapExtractAdapter(headers));
+
+        spanBuilder = tracer.BuildSpan(operationName);
+        if (parentSpanCtx != null)
+        {
+            spanBuilder = spanBuilder.AsChildOf(parentSpanCtx);
+        }
+    }
+    catch (Exception)
+    {
+        spanBuilder = tracer.BuildSpan(operationName);
+    }
+
+    // TODO could add more tags like http.url
+    return spanBuilder.WithTag(Tags.SpanKind, Tags.SpanKindServer).StartActive(true);
+}
+```
+
+The logic here is similar to the client side instrumentation, except that we are using `_tracer.Extract`
+and tagging the span as `span.kind=server`. We are extracting the headers via `TextMapExtractAdapter`.
+
+Now change the `/api/format/helloString` handler method to use `StartServerSpan`:
+
+```csharp
+[HttpGet("{helloString}", Name = "GetFormat")]
+public string Get(string helloString)
+{
+    var headers = Request.Headers.ToDictionary(k => k.Key, v => v.Value.First());
+    using (var scope = StartServerSpan(_tracer, headers, "format-controller"))
```

Comment:
s/format-controller/format-string/

Reply:
`format-string` is used on the client side and `format-controller` on the server side. If I change it on the server side, you would get two spans (one root, one child) with the same operation name which does not help understanding.

Reply:
that's actually very common and expected. In Jaeger UI you'd be able to tell them apart by different service names.

Reply:
You are right

Reply:
Just thought about changing it, but I don’t have docker installed here at home. So I can‘t recreate the [image](https://github.com/saschaholesch/opentracing-tutorial/blob/ed4e5b0e9fbd5adad798f7e053f0fe71b11bed52/csharp/src/lesson03/trace-server-unfinished.png) :( so either we keep it for now, or I will change it next Monday when I‘m back in the office. My machine at home does  not support Hyper-V.